bve-render/src/lib.rs

```diff
@@ -0,0 +1,378 @@
+
+// Rust warnings
+#![warn(unused)]
+#![deny(future_incompatible)]
+#![deny(nonstandard_style)]
+#![deny(rust_2018_idioms)]
+// Rustdoc Warnings
+#![deny(intra_doc_link_resolution_failure)]
+// Clippy warnings
+#![warn(clippy::cargo)]
+#![warn(clippy::nursery)]
+#![warn(clippy::pedantic)]
+#![warn(clippy::restriction)]
+// Annoying regular clippy warnings
+#![allow(clippy::cast_lossless)] // Annoying
+#![allow(clippy::cast_sign_loss)] // Annoying
+#![allow(clippy::cast_precision_loss)] // Annoying
+#![allow(clippy::cast_possible_truncation)] // Annoying
+#![allow(clippy::cognitive_complexity)] // This is dumb
+#![allow(clippy::too_many_lines)] // This is also dumb
+// Annoying/irrelevant clippy Restrictions
+#![allow(clippy::as_conversions)]
+#![allow(clippy::decimal_literal_representation)]
+#![allow(clippy::else_if_without_else)]
+#![allow(clippy::fallible_impl_from)] // This fails horribly when you try to panic in a macro inside a From impl
+#![allow(clippy::float_arithmetic)]
+#![allow(clippy::float_cmp)]
+#![allow(clippy::float_cmp_const)]
+#![allow(clippy::implicit_return)]
+#![allow(clippy::indexing_slicing)]
+#![allow(clippy::integer_arithmetic)]
+#![allow(clippy::integer_division)]
+#![allow(clippy::let_underscore_must_use)]
+#![allow(clippy::match_bool)] // prettier
+#![allow(clippy::missing_docs_in_private_items)]
+#![allow(clippy::missing_inline_in_public_items)]
+#![allow(clippy::module_name_repetitions)]
+#![allow(clippy::multiple_crate_versions)] // Cargo deny's job
+#![allow(clippy::multiple_inherent_impl)]
+#![allow(clippy::non_ascii_literal)]
+#![allow(clippy::option_expect_used)]
+#![allow(clippy::panic)]
+#![allow(clippy::result_expect_used)]
+#![allow(clippy::similar_names)]
+#![allow(clippy::shadow_reuse)]
+#![allow(clippy::shadow_same)]
+#![allow(clippy::string_add)]
+#![allow(clippy::unreachable)]
+#![allow(clippy::wildcard_enum_match_arm)]
+
+pub use crate::{object::ObjectHandle, render::MSAASetting, texture::Texture};
+use bve::load::mesh::Vertex as MeshVertex;
+use cgmath::{Matrix4, Vector2, Vector3};
+use image::RgbaImage;
+use indexmap::map::IndexMap;
+use itertools::Itertools;
+use num_traits::{ToPrimitive, Zero};
+use std::io;
+use wgpu::*;
+use winit::{dpi::PhysicalSize, window::Window};
+use zerocopy::{AsBytes, FromBytes};
+
+macro_rules! include_shader {
+    (vert $name:literal) => {
+        include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/shaders/", $name, ".vs.spv"));
+    };
+    (geo $name:literal) => {
+        include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/shaders/", $name, ".gs.spv"));
+    };
+    (frag $name:literal) => {
+        include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/shaders/", $name, ".fs.spv"));
+    };
+    (comp $name:literal) => {
+        include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/shaders/", $name, ".cs.spv"));
+    };
+}
+
+mod camera;
+mod compute;
+mod object;
+mod render;
+mod texture;
+
+pub const OPENGL_TO_WGPU_MATRIX: Matrix4<f32> = Matrix4::new(
```

Comment:
Yeah, mainly because `cgmath::perspective` assumes OpenGL, so I can't turn this into `DIRECTX_TO_WGPU_MATRIX`. This is one of those nice things that will come later.

Reply:
Alright